src/Expander.hs

```diff
@@ -340,39 +343,41 @@ visibleBindings = do
 
 
 
-allMatchingBindings :: Text -> ScopeSet -> Expand [(ScopeSet, Binding)]
+allMatchingBindings :: Text -> ScopeSet -> Expand (Seq (ScopeSet, Binding))
 allMatchingBindings x scs = do
-  namesMatch <- view (at x . non []) <$> visibleBindings
+  namesMatch <- view (at x . non mempty) <$> visibleBindings
   p <- currentPhase
-  let scopesMatch =
-        [ (scopes, b)
-        | (scopes, b, _) <- namesMatch
-        , ScopeSet.isSubsetOf p scopes scs
-        ]
+  let scopesMatch = fmap (\(s,b,_) -> (s,b))
+        . Seq.filter (\(scopes, _, _) -> ScopeSet.isSubsetOf p scopes scs)
+        $ namesMatch
   return scopesMatch
 
 
 
-checkUnambiguous :: ScopeSet -> [ScopeSet] -> Ident -> Expand ()
+checkUnambiguous :: ScopeSet -> (Seq ScopeSet) -> Ident -> Expand ()
 checkUnambiguous best candidates blame =
   do p <- currentPhase
      let bestSize = ScopeSet.size p best
-     let candidateSizes = map (ScopeSet.size p) (nub candidates)
-     if length (filter (== bestSize) candidateSizes) > 1
-       then throwError (Ambiguous p blame candidates)
-       else return ()
+     let candidateSizes = Seq.fromList $ fmap (ScopeSet.size p) (nub $ toList candidates)
+     when (Seq.length (Seq.filter (== bestSize) candidateSizes) > 1)
```

Comment:
If the list is small enough that you're not concerned about the performance of `Data.List.nub`, it's probably not worth converting it back to `Seq` :)

This list should usually have length 1 or 2, as it is the list of shadowed variables with the same name and users don't typically shadow a variable which has already been shadowed. That being said, some macros can and do generate nested binders which all have the same dummy variable name, relying on the ScopeSet to disambiguate, so in that case the list could get longer, maybe a dozen elements.

Reply:
ok great then I can revert this one

Reply:
see 2e45103d571a88c9ceaf247a8b715d751c0f3f86